File: lingting-spring-redis/src/main/java/live/lingting/spring/redis/cache/CacheKeyGenerator.java
package live.lingting.spring.redis.cache;

import live.lingting.framework.util.CollectionUtils;
import live.lingting.framework.util.StringUtils;
import live.lingting.spring.util.SpelUtils;
import org.springframework.expression.spel.support.StandardEvaluationContext;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * @author lingting 2024-04-18 10:04
 */
public class CacheKeyGenerator {

	protected final String delimiter;

	protected final StandardEvaluationContext context;

	public CacheKeyGenerator(String delimiter, Object target, Method method, Object[] arguments) {
		this.delimiter = delimiter;
		this.context = SpelUtils.getSpelContext(target, method, arguments);
	}

	public List<String> resolve(String spel) {
		if (!StringUtils.hasText(spel)) {
			return Collections.emptyList();
		}

		Object obj = SpelUtils.parseValue(context, spel);
		return CollectionUtils.multiToList(obj).stream().map(o -> o == null ? "" : o.toString()).toList();
	}

	public String join(String key, String spel) {
		StringBuilder builder = new StringBuilder(key);

		List<String> list = resolve(spel);

		if (!CollectionUtils.isEmpty(list)) {
			for (String s : list) {
				builder.append(delimiter).append(s);
			}
		}

		return builder.toString();
	}

	public List<String> multi(String key, String spel) {
		List<String> list = resolve(spel);
		if (!CollectionUtils.isEmpty(list)) {
			return Collections.singletonList(key);
		}

		List<String> strings = new ArrayList<>(list.size());

		for (String s : list) {
			strings.add("%s%s%s".formatted(key, delimiter, s));
		}

		return strings;
	}

	public String cached(Cached cached) {
		if (cached == null) {
			return "";
		}
		return join(cached.key(), cached.keyJoint());
	}

	public List<String> cacheClear(CacheClear clear) {
		if (clear == null) {
			return Collections.emptyList();
		}
		if (clear.multi()) {
			return multi(clear.key(), clear.keyJoint());
		}
		return Collections.singletonList(join(clear.key(), clear.keyJoint()));
	}

	public List<String> cacheClear(CacheClear clear, CacheBatchClear batchClear) {
		List<String> strings = new ArrayList<>(cacheClear(clear));
		for (CacheClear cacheClear : batchClear.value()) {
			strings.addAll(cacheClear(cacheClear));
		}
		return strings;
	}

}

File: lingting-spring-redis/src/main/kotlin/live/lingting/spring/redis/cache/CacheKeyGenerator.kt
package live.lingting.spring.redis.cache

import live.lingting.framework.util.CollectionUtils.multiToList
import live.lingting.framework.util.StringUtils.hasText
import live.lingting.spring.util.SpelUtils
import org.springframework.expression.spel.support.StandardEvaluationContext
import java.lang.reflect.Method

/**
 * @author lingting 2024-04-18 10:04
 */
class CacheKeyGenerator(
    val delimiter: String,
    target: Any,
    method: Method,
    arguments: Array<Any>
) {
    val context: StandardEvaluationContext = SpelUtils.getSpelContext(target, method, arguments)

    fun resolve(spel: String): List<String> {
        if (!hasText(spel)) {
            return mutableListOf()
        }

        val obj = SpelUtils.parseValue(context, spel)
        return multiToList(obj).stream().map { it?.toString() ?: "" }.toList()
    }

    fun join(key: String, spel: String): String {
        val builder = StringBuilder(key)

        val list = resolve(spel)

        if (!list.isEmpty()) {
            for (s in list) {
                builder.append(delimiter).append(s)
            }
        }

        return builder.toString()
    }

    fun multi(key: String, spel: String): List<String> {
        val list = resolve(spel)
        if (list.isEmpty()) {
            return mutableListOf(key)
        }

        val strings = ArrayList<String>(list.size)

        for (s in list) {
            strings.add("$key$delimiter$s")
        }

        return strings.distinct()
    }

    fun cached(a: Cached?): List<String> {
        if (a == null) {
            return mutableListOf()
        }
        if (a.keyMulti) {
            return multi(a.key, a.keyJoint)
        }
        return mutableListOf(join(a.key, a.keyJoint))
    }

    fun cached(a: Cached?, batch: CachedBatch?): List<String> {
        val keys = mutableListOf<String>()
        keys.addAll(cached(a))
        batch?.let {
            for (ba in it.value) {
                keys.addAll(cached(ba))
            }
        }
        return keys.distinct()
    }

    fun cacheClear(a: CacheClear?): List<String> {
        if (a == null) {
            return mutableListOf()
        }
        if (a.keyMulti) {
            return multi(a.key, a.keyJoint)
        }
        return mutableListOf(join(a.key, a.keyJoint))
    }

    fun cacheClear(a: CacheClear?, batch: CacheClearBatch?): List<String> {
        val keys = mutableListOf<String>()
        keys.addAll(cacheClear(a))
        batch?.let {
            for (ba in it.value) {
                keys.addAll(cacheClear(ba))
            }
        }
        return keys.distinct()
    }
}
